[NFC][LoopVectorize] Introduce new getEstimatedRuntimeVF function

david-arm opened the pull request:
There are lots of places where we try to estimate the runtime
vectorisation factor based on the getVScaleForTuning TTI hook.
I've added a new getEstimatedRuntimeVF function and taught
several places in the vectoriser to use this new function.

hazzlim commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4976,11 +4973,7 @@ LoopVectorizationCostModel::selectInterleaveCount(ElementCount VF,
-      EstimatedVF *= *VScale;
-  }
+  unsigned EstimatedVF = getEstimatedRuntimeVF(TheLoop, TTI, VF);
   assert(EstimatedVF >= 1 && "Estimated VF shouldn't be less than 1");
```
Does it make any sense to move this assert into `getEstimatedRuntimeVF` ?

david-arm replied:
It's a good question. It looks like it was really just asserting that `getVScaleForTuning` returned a non-zero value. Maybe it is worth moving into `getEstimatedRuntimeVF`.

hazzlim reviewed (approved):
LGTM! :)

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4300,6 +4300,17 @@ getVScaleForTuning(const Loop *L, const TargetTransformInfo &TTI) {
   return TTI.getVScaleForTuning();
 }
 
+static unsigned getEstimatedRuntimeVF(const Loop *L,
```
Would be gerat if you could add a comment here to document exactly what we mean by estimated runtime VF here

fhahn reviewed (commented):
LGTM, thanks!

david-arm committed:
[NFC][LoopVectorize] Introduce new getEstimatedRuntimeVF function

There are lots of places where we try to estimate the runtime
vectorisation factor based on the getVScaleForTuning TTI hook.
I've added a new getEstimatedRuntimeVF function and taught
several places in the vectoriser to use this new function.

david-arm commented:
Rebase + added a TODO in isEpilogueVectorizationProfitable

hazzlim commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4291,6 +4291,21 @@ getVScaleForTuning(const Loop *L, const TargetTransformInfo &TTI) {
   return TTI.getVScaleForTuning();
 }
 
+/// This functions attempts to return a value that represents the vectorization
```
Just a small nit:
```suggestion
/// This function attempts to return a value that represents the vectorization
```

david-arm commented:
Rebase. Updated `LoopVectorizationPlanner::cost` to use new function.

david-arm committed:
Merge branch 'main' into estimate_vf
